sui_programmability/examples/games_with_chance/sources/satoshi_flip.move

```diff
@@ -0,0 +1,300 @@
+    const EpochsCancelAfter: u64 = 7;
+
+    // errors
+    const EStakeTooHigh: u64 = 0; // this should hold stake <= max bet
+    const EStakeTooLow: u64 = 1; // this should hold stake >= min bet
+    const EGuessNot1Or0: u64 = 2; // guess should be either 1 or 0
+    const EHashAndSecretDontMatch: u64 = 3; // foul play or mistake, house provided wrong secret
+    const EMinBetTooHigh: u64 = 4; // house's min bet must be lower or equal to max bet
+    const EGameAlreadyEnded: u64 = 5; // Too late to cancel a game that has been settled
+    const ENotEnoughEpochsPassedToCancel: u64 = 6; // Can only cancel after above EpochsCancelAfter epochs, after the bet was placed
+    const ENotAllowedToEndGame: u64 = 7; // Only house may end a game
+    const ECannotCancelBeforeBetting: u64 = 8; // Only a game that has received a bet may be cancelled
+    const EHouseCoinNotEnough: u64 = 9; // House provided coin with insuficient balance to cover max bet
+    const EPlayerCoinNotEnoughBalance: u64 = 10; // Player provided a coin with insufficient balance to cover his stake
+    const EGameNotEnded: u64 = 11;
+    const EAlreadyAcceptedBet: u64 = 12;
+    const ESecretIsEmpty: u64 = 13; // secret can't be an empty string
+    const EZeroMinBet: u64 = 14;
+
+    // structs
+
+    /// Player's data, the stake of the wager and his guess
+    struct BetData has store {
+        stake: Balance<SUI>,
+        guess: u8,
+    }
+
+    /// House's data, min_bet is the minimum amount accepted for a bet
+    /// max_bet is the maximum amount accepted for a bet
+    struct HouseData has store {
+        house_balance: Balance<SUI>,
+        min_bet: u64,
+        max_bet: u64
+    }
+
+    /// Outcome will hold end game data, the secret the house picked, the player's guess and if the player won
+    /// This way anyone can check if the winner was correctly determined
+    struct Outcome has store {
+        secret: vector<u8>,
+        guess: u8,
+        player_won: bool
+    }
+
+    /// Each Game is a shared object, the address who created it must also end it.
+    /// The player can only be a single address for each game
+    ///
+    ///  @ownership: Shared
+    ///
+
+    struct Game has key {
+        id: UID,
+        bet_placed_epoch: u64,
+        hashed_secret: Sha3256Digest,
+        house: address,
+        player: Option<address>,
+        house_data: Option<HouseData>,
+        bet_data: Option<BetData>,
+        outcome: Option<Outcome>
+    }
+
+    // fun(ctions)!
+
+    /// creates a new game and makes it a transfered object
+    /// a user who wants to become house should call this function
+    public entry fun start_game(hash: vector<u8>, house_coin: Coin<SUI>, max_bet: u64, min_bet: u64, ctx: &mut TxContext) {
+        assert!(min_bet > 0, EZeroMinBet);
+        assert!(max_bet >= min_bet, EMinBetTooHigh);
+        assert!(coin::value(&house_coin) >= max_bet, EHouseCoinNotEnough);
```

Comment:
Add a comment here that we accept `house_coin_balance >= max_bet`, although in practice they should actually match. We allow this just because in Sui everything is an object (and thus to avoid splitting coins), but this contract's logic ensures that any surplus will return to the house, so it's safe to accept `house_coin_balance >= max_bet`.
The recommendation is (of course) for the house to pick a coin as close to max_bet as possible to avoid locking higher balances.